### PR TITLE
wasm64 support

### PR DESCRIPTION
**Pull request by devsnek**
There is still some upstream llvm work needed before this can land.

### CODE REVIEW

**Comment by devsnek** [comment]
@nagisa not blocked anymore :tada:

**Comment by oli-obk** [comment]
r? @nagisa

I won't be able to review in the next 3 months

File: compiler/rustc_codegen_llvm/src/attributes.rs [hunk 1]
```diff
@@ -317,7 +317,7 @@ pub fn from_fn_attrs(cx: &CodegenCx<'ll, 'tcx>, llfn: &'ll Value, instance: ty::
     // Note that currently the `wasm-import-module` doesn't do anything, but
     // eventually LLVM 7 should read this and ferry the appropriate import
     // module to the output file.
-    if cx.tcx.sess.target.arch == "wasm32" {
+    if cx.tcx.sess.target.arch.starts_with("wasm") {
```
Comment (nagisa) [inline compiler/rustc_codegen_llvm/src/attributes.rs:320]: This should probably add a target property `is_like_wasm` or similar, much like e.g. `is_like_emscripten`, `is_like_windows` and similar.

**Comment by nagisa** [comment]
This overall LGTM, but the test runner needs to be adjusted to understand wasm32 vs wasm64.

**Comment by devsnek** [comment]
@nagisa should be good to go now.

**Comment by nagisa** [comment]
@bors r+

(cc @alexcrichton as you're heavily involved with wasm-related support)

**Comment by devsnek** [comment]
I think it would be worth noting `cfg(wasm)` > `cfg(target_arch = "wasm32")` in the next changelog, even if there is still a ways to go before wasm64 is fully working.

**Comment by nagisa** [comment]
cc @rust-lang/lang I guess I should also ask if we're comfortable adding such a cfg. This would be the first top level one that is set based on the architecture being targeted – we don't have one that e.g. unifies matching `x86_64` vs `x86_32`.

**Comment by Havvy** [comment]
This also needs to be documented on the reference should T-lang's RFC review pass.

**Pull request merged**